// @arch-ui/pagination/dist/pagination.esm.js
import _extends from '@babel/runtime/helpers/esm/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import { jsx } from '@emotion/core';
import { createElement, Component, useState, useEffect } from 'react';
import { ChevronLeftIcon, ListOrderedIcon, ChevronRightIcon } from '@arch-ui/icons';
import { FlexGroup } from '@arch-ui/layout';
import { LoadingSpinner } from '@arch-ui/loading';
import { Button } from '@arch-ui/button';

const Page = props => {
  const handleClick = () => {
    if (props.onClick) {
      props.onClick(props.value);
    }
  };

  return createElement(Button, _extends({}, props, {
    onClick: handleClick
  }));
};

function ariaPageLabelFn(page) {
  return "Go to page ".concat(page);
}

const PageChildren = (_ref) => {
  let {
    page,
    isLoading,
    isSelected
  } = _ref;
  const [shouldShowLoading, setShouldShowLoading] = useState(false);
  useEffect(() => {
    if (isLoading && isSelected) {
      const id = setTimeout(() => {
        setShouldShowLoading(true);
      }, 200);
      return () => {
        clearTimeout(id);
        setShouldShowLoading(false);
      };
    }
  }, [page, isLoading, isSelected]);
  return shouldShowLoading ? jsx("div", {
    css: {
      height: 19
    }
  }, jsx(LoadingSpinner, null)) : jsx("span", null, page);
};

class Pagination extends Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      allPagesVisible: false
    });

    _defineProperty(this, "toggleAllPages", () => {
      this.setState(state => ({
        allPagesVisible: !state.allPagesVisible
      }));
    });
  }

  renderPages() {
    let {
      ariaPageLabel,
      currentPage,
      limit,
      pageSize,
      total
    } = this.props;
    if (total <= pageSize) return [];
    let pages = [];
    let totalPages = Math.ceil(total / pageSize);
    let minPage = 1;
    let maxPage = totalPages;
    const moreCharacter = jsx("span", null, "\u2026");

    if (limit && limit < totalPages) {
      let rightLimit = Math.floor(limit / 2);
      let leftLimit = rightLimit + limit % 2 - 1;
      minPage = currentPage - leftLimit;
      maxPage = currentPage + rightLimit;

      if (minPage < 1) {
        maxPage = limit;
        minPage = 1;
      }

      if (maxPage > totalPages) {
        minPage = totalPages - limit + 1;
        maxPage = totalPages;
      }
    }

    const onChange = page => {
      if (this.props.onChange) {
        this.setState({
          allPagesVisible: false
        });
        this.props.onChange(page, {
          pageSize,
          total,
          minPage,
          maxPage
        });
      }
    }; // go to first


    if (minPage > 1) {
      pages.push(jsx(Page, {
        "aria-label": ariaPageLabel(1),
        key: "page_start",
        onClick: onChange,
        value: 1
      }, moreCharacter));
    } // loop over range


    for (let page = minPage; page <= maxPage; page++) {
      const isSelected = page === currentPage;
      pages.push(jsx(Page, {
        "aria-label": ariaPageLabel(page),
        "aria-current": isSelected ? 'page' : null,
        key: "page_".concat(page),
        isSelected: isSelected,
        onClick: onChange,
        value: page
      }, jsx(PageChildren, {
        isLoading: this.props.isLoading,
        page: page,
        isSelected: isSelected
      })));
    } // go to last


    if (maxPage < totalPages) {
      pages.push(jsx(Page, {
        "aria-label": ariaPageLabel(totalPages),
        key: "page_end",
        onClick: onChange,
        value: totalPages
      }, moreCharacter));
    } // return pages;


    return [jsx(Page, {
      "aria-label": "Go to previous page",
      key: "page_prev",
      onClick: onChange,
      value: currentPage - 1,
      isDisabled: currentPage === 1
    }, jsx(ChevronLeftIcon, null)), this.state.allPagesVisible ? pages : jsx(Page, {
      "aria-label": "Click to show all pages",
      key: "page_dot",
      onClick: this.toggleAllPages,
      id: "ks-pagination-show-pages",
      value: 1 // needs value for flow...

    }, jsx(ListOrderedIcon, null)), jsx(Page, {
      "aria-label": "Go to next page",
      key: "page_next",
      onClick: onChange,
      value: currentPage + 1,
      isDisabled: currentPage === totalPages
    }, jsx(ChevronRightIcon, null))];
  }

  render() {
    // strip props to get `rest` attributes; things id, className etc.
    const _this$props = this.props,
          rest = _objectWithoutProperties(_this$props, ["ariaPageLabel", "currentPage", "isLoading", "limit", "onChange", "pageSize", "total"]);

    return jsx(FlexGroup, _extends({
      as: "nav",
      align: "center",
      "aria-label": "Pagination",
      isContiguous: true,
      isInline: true
    }, rest), this.renderPages());
  }

}

_defineProperty(Pagination, "defaultProps", {
  ariaPageLabel: ariaPageLabelFn,
  currentPage: 1,
  limit: 5
});

export { Pagination };
