// @arch-ui/typography/dist/typography.cjs.prod.js
"use strict";

function _interopDefault(ex) {
  return ex && "object" == typeof ex && "default" in ex ? ex.default : ex;
}

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var _defineProperty = _interopDefault(require("@babel/runtime/helpers/defineProperty")), _extends = _interopDefault(require("@babel/runtime/helpers/extends")), _objectWithoutProperties = _interopDefault(require("@babel/runtime/helpers/objectWithoutProperties")), core = require("@emotion/core"), styled = _interopDefault(require("@emotion/styled")), theme = require("@arch-ui/theme");

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter((function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    }))), keys.push.apply(keys, symbols);
  }
  return keys;
}

function _objectSpread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = null != arguments[i] ? arguments[i] : {};
    i % 2 ? ownKeys(Object(source), !0).forEach((function(key) {
      _defineProperty(target, key, source[key]);
    })) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach((function(key) {
      Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
    }));
  }
  return target;
}

const truncate = {
  minWidth: 0,
  overflow: "hidden",
  textOverflow: "ellipsis",
  whiteSpace: "nowrap"
}, SubtleText = styled.span({
  color: theme.colors.N60
}), H1 = styled.h1({
  fontSize: 28,
  fontWeight: 300,
  margin: "24px 0"
}), PageTitle = styled.h1({
  fontSize: "2.4rem",
  fontWeight: 900,
  marginBottom: "0.5em",
  marginTop: "1.0rem"
}), Truncate = _ref => {
  let {as: Tag} = _ref, props = _objectWithoutProperties(_ref, [ "as" ]);
  return core.jsx(Tag, _extends({
    css: truncate
  }, props));
};

Truncate.defaultProps = {
  as: "div"
};

const Title = _ref2 => {
  let {as: Tag, crop: crop, margin: margin} = _ref2, props = _objectWithoutProperties(_ref2, [ "as", "crop", "margin" ]);
  const gutter = 3 * theme.gridSize, offset = {
    none: {
      margin: 0
    },
    both: {
      marginBottom: gutter,
      marginTop: gutter
    },
    bottom: {
      marginBottom: gutter,
      marginTop: 0
    },
    top: {
      marginBottom: 0,
      marginTop: gutter
    }
  }[margin], cropStyles = crop ? truncate : null;
  return core.jsx(Tag, _extends({
    css: _objectSpread({
      fontSize: 18,
      fontWeight: 600,
      whiteSpace: "nowrap"
    }, cropStyles, {}, offset)
  }, props));
};

Title.defaultProps = {
  as: "h3",
  crop: !1,
  margin: "none"
};

const Kbd = styled.kbd({
  backgroundColor: theme.colors.N05,
  border: "1px solid ".concat(theme.colors.N20),
  borderRadius: 3,
  boxShadow: "0 1px 1px rgba(0, 0, 0, 0.12), 0 2px 0 0 rgba(255, 255, 255, 0.7) inset",
  display: "inline-block",
  fontFamily: "Monaco, monospace",
  fontSize: "0.85em",
  fontWeight: "bold",
  lineHeight: "inherit",
  padding: "1px 5px",
  position: "relative",
  top: "-1px",
  verticalAlign: "middle",
  whiteSpace: "nowrap"
}), A11yText = _ref3 => {
  let {tag: Tag} = _ref3, props = _objectWithoutProperties(_ref3, [ "tag" ]);
  return core.jsx(Tag, _extends({
    css: {
      border: 0,
      clip: "rect(1px, 1px, 1px, 1px)",
      height: 1,
      overflow: "hidden",
      padding: 0,
      position: "absolute",
      whiteSpace: "nowrap",
      width: 1
    }
  }, props));
};

A11yText.defaultProps = {
  tag: "span"
}, exports.A11yText = A11yText, exports.H1 = H1, exports.Kbd = Kbd, exports.PageTitle = PageTitle, 
exports.SubtleText = SubtleText, exports.Title = Title, exports.Truncate = Truncate;
