// react-uid/dist/es5/Component.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var uid_1 = require("./uid");
exports.createSource = function (prefix) {
    if (prefix === void 0) { prefix = ''; }
    return ({
        value: 1,
        prefix: prefix,
        uid: uid_1.generateUID()
    });
};
exports.counter = exports.createSource();
exports.source = React.createContext(exports.createSource());
exports.lastUsed = React.createContext("");
exports.getId = function (source) { return source.value++; };
exports.getPrefix = function (source) { return source ? source.prefix : ''; };
var UID = (function (_super) {
    __extends(UID, _super);
    function UID() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            quartz: _this.props.idSource || exports.counter,
            prefix: exports.getPrefix(_this.props.idSource),
            id: exports.getId(_this.props.idSource || exports.counter)
        };
        _this.uid = function (item) { return _this.prefixId(_this.state.quartz.uid(item)); };
        return _this;
    }
    UID.prototype.prefixId = function (id) {
        var uid = (this.state.prefix + id);
        return String(this.props.name ? this.props.name(uid) : uid);
    };
    UID.prototype.render = function () {
        var children = this.props.children;
        var id = this.state.id;
        return children(this.prefixId(id), this.uid);
    };
    return UID;
}(React.Component));
exports.UID = UID;
exports.UIDReset = function (_a) {
    var children = _a.children;
    return (React.createElement(exports.source.Provider, { value: exports.createSource() }, children));
};
exports.UIDFork = function (_a) {
    var children = _a.children;
    return (React.createElement(exports.UIDConsumer, null, function (id) { return (React.createElement(exports.source.Provider, { value: exports.createSource(id + '-') }, children)); }));
};
exports.UIDConsumer = function (_a) {
    var name = _a.name, children = _a.children;
    return (React.createElement(exports.source.Consumer, null, function (value) { return (React.createElement(UID, { name: name, idSource: value, children: children })); }));
};
