// @keystonejs/field-content/dist/image-812ee484.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var _objectWithoutProperties = _interopDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var core = require('@emotion/core');

var _extends = _interopDefault(require("@babel/runtime/helpers/extends"));

var React = require('react');

var React__default = _interopDefault(React);

var reactPopper = require('react-popper');

var index = require('./index-eb1bb328.cjs.js');

let PopperRender = React.forwardRef((_ref, ref) => {
  let {
    scheduleUpdate,
    alignment,
    isFocused,
    style,
    onAlignmentChange
  } = _ref;
  React.useLayoutEffect(scheduleUpdate, [alignment]);
  return core.jsx("div", {
    ref: ref,
    css: {
      display: isFocused ? 'block' : 'none',
      backgroundColor: 'black',
      padding: 8
    },
    style: style
  }, ['left', 'center', 'right'].map(align => {
    return core.jsx("button", {
      type: "button",
      key: align,
      onMouseDown: event => {
        // so that the image block doesn't get deselected
        event.preventDefault();
      },
      onClick: () => {
        onAlignmentChange(align);
      }
    }, align);
  }));
});
let popperModifiers = {
  flip: {
    enabled: false
  },
  hide: {
    enabled: false
  },
  preventOverflow: {
    enabled: false
  }
};

let Image = _ref2 => {
  let {
    alignment,
    attributes,
    isFocused,
    src,
    onAlignmentChange
  } = _ref2,
      props = _objectWithoutProperties(_ref2, ["alignment", "attributes", "isFocused", "src", "onAlignmentChange"]);

  let [referenceElement, setReferenceElement] = index.useStateWithEqualityCheck(null);
  return core.jsx("div", null, core.jsx("img", _extends({}, props, attributes, {
    src: src,
    ref: setReferenceElement,
    css: {
      width: '100%',
      outline: isFocused ? 'auto' : null
    }
  })), core.jsx(reactPopper.Popper, {
    modifiers: popperModifiers,
    placement: "top",
    referenceElement: referenceElement
  }, _ref3 => {
    let {
      style,
      ref,
      scheduleUpdate
    } = _ref3;
    return core.jsx(PopperRender, {
      scheduleUpdate,
      alignment,
      ref,
      isFocused,
      style,
      onAlignmentChange
    });
  }));
};
/** @jsx jsx */


let type = 'image';
let ImageAlignmentContext = React.createContext({
  aligment: '',

  onAlignmentChange() {}

});

function Node(props) {
  let {
    data
  } = props.node;
  let {
    aligment,
    onAlignmentChange
  } = React.useContext(ImageAlignmentContext);
  return core.jsx(Image, {
    alignment: aligment,
    attributes: props.attributes,
    isFocused: props.isFocused,
    src: data.get('src'),
    onAlignmentChange: onAlignmentChange
  });
}

let getSchema = () => ({
  isVoid: true
});

exports.ImageAlignmentContext = ImageAlignmentContext;
exports.Node = Node;
exports.getSchema = getSchema;
exports.type = type;
