// @keystonejs/field-content/dist/unordered-list-c0b3d736.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

require("@babel/runtime/helpers/objectWithoutProperties");

require('@emotion/core');

require("@babel/runtime/helpers/extends");

var utils = require('./utils-29e2a896.cjs.js');

var icons = require('@arch-ui/icons');

require('@arch-ui/theme');

require('@arch-ui/color-utils');

require('@arch-ui/tooltip');

require('@arch-ui/typography');

var toolbarComponents = require('./toolbar-components-f0fc99c4.cjs.js');

var React = require('react');

var React__default = _interopDefault(React);

var paragraph = require('./paragraph-d4d0e6d8.cjs.js');

var listItem = require('./list-item-c004e639.cjs.js');

let handleListButtonClick = (editor, editorState, type) => {
  let isListItem = utils.hasBlock(editorState, listItem.type);
  let isOrderedList = utils.hasAncestorBlock(editorState, type);
  let otherListType = type === 'ordered-list' ? 'unordered-list' : 'ordered-list';

  if (isListItem && isOrderedList) {
    editor.setBlocks(paragraph.type);
    editor.unwrapBlock(type);
  } else if (isListItem) {
    editor.unwrapBlock(otherListType);
    editor.wrapBlock(type);
  } else {
    editor.setBlocks(listItem.type).wrapBlock(type);
  }

  editor.focus();
};

let type = 'unordered-list';

function ToolbarElement(_ref) {
  let {
    editor,
    editorState
  } = _ref;
  return React.createElement(toolbarComponents.ToolbarButton, {
    label: "Unordered List",
    icon: React.createElement(icons.ListUnorderedIcon, null),
    isActive: utils.hasAncestorBlock(editorState, type),
    onClick: () => {
      handleListButtonClick(editor, editorState, type);
    }
  });
}

let getPlugins = () => [{
  onKeyDown(event, editor, next) {
    // make it so when you press enter in an empty list item,
    // the block type will change to a paragraph
    if (event.keyCode === 13 && utils.hasAncestorBlock(editor.value, type) && editor.value.focusText.text === '') {
      editor.setBlocks(paragraph.type).unwrapBlock(type);
      return;
    }

    next();
  }

}];

function Node(_ref2) {
  let {
    attributes,
    children
  } = _ref2;
  return React.createElement("ul", attributes, children);
}

const getSchema = () => ({
  nodes: [{
    match: {
      type: listItem.type
    },
    min: 0
  }],

  normalize(editor, error) {
    switch (error.code) {
      case 'child_type_invalid':
        {
          editor.unwrapBlockByKey(error.node.key, type);
          return;
        }
    }
  }

});

exports.Node = Node;
exports.ToolbarElement = ToolbarElement;
exports.getPlugins = getPlugins;
exports.getSchema = getSchema;
exports.type = type;
