// @arch-ui/card/dist/card.cjs.prod.js
"use strict";

function _interopDefault(ex) {
  return ex && "object" == typeof ex && "default" in ex ? ex.default : ex;
}

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var _extends = _interopDefault(require("@babel/runtime/helpers/extends")), _objectWithoutProperties = _interopDefault(require("@babel/runtime/helpers/objectWithoutProperties")), core = require("@emotion/core"), theme = require("@arch-ui/theme");

const Card = _ref => {
  let {as: Tag, elevation: elevation, isInteractive: isInteractive, isPadded: isPadded} = _ref, props = _objectWithoutProperties(_ref, [ "as", "elevation", "isInteractive", "isPadded" ]);
  const shadow = theme.shadows[elevation];
  return core.jsx(Tag, _extends({
    css: {
      background: "white",
      borderRadius: theme.borderRadius,
      boxShadow: shadow,
      padding: isPadded ? 2 * theme.gridSize : null,
      transition: "box-shadow 111ms ease-out, transform 111ms ease-out",
      ":hover": isInteractive ? {
        boxShadow: theme.shadows[elevation + 1],
        transform: "translateY(-2px)"
      } : null,
      ":active": isInteractive ? {
        boxShadow: shadow,
        transform: "none"
      } : null
    }
  }, props));
};

Card.defaultProps = {
  as: "div",
  elevation: 0,
  isPadded: !0
};

const Canvas = _ref2 => {
  let {as: Tag, isPadded: isPadded} = _ref2, props = _objectWithoutProperties(_ref2, [ "as", "isPadded" ]);
  return core.jsx(Tag, _extends({
    css: {
      background: "rgba(9, 30, 66, 0.04)",
      borderRadius: 2 * theme.borderRadius,
      padding: isPadded ? theme.gridSize : null
    }
  }, props));
};

Canvas.defaultProps = {
  as: "div",
  isPadded: !0
}, exports.Canvas = Canvas, exports.Card = Card;
