// @arch-ui/theme/dist/theme.cjs.prod.js
"use strict";

function _interopDefault(ex) {
  return ex && "object" == typeof ex && "default" in ex ? ex.default : ex;
}

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var _defineProperty = _interopDefault(require("@babel/runtime/helpers/defineProperty")), colorUtils = require("@arch-ui/color-utils");

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter((function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    }))), keys.push.apply(keys, symbols);
  }
  return keys;
}

function _objectSpread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = null != arguments[i] ? arguments[i] : {};
    i % 2 ? ownKeys(Object(source), !0).forEach((function(key) {
      _defineProperty(target, key, source[key]);
    })) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach((function(key) {
      Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
    }));
  }
  return target;
}

const intent = {
  create: "#34c240",
  danger: "#d64242",
  info: "#0090e0",
  primary: "#2684FF",
  warning: "#fa9f47"
}, neutrals = {
  N05: "#F4F5F7",
  N10: "#EBECF0",
  N15: "#DFE1E5",
  N20: "#C1C7D0",
  N30: "#A5ADBA",
  N40: "#97A0AF",
  N50: "#7A869A",
  N60: "#6C798F",
  N70: "#42526E",
  N80: "#253858",
  N90: "#172B4D",
  N100: "#091E42"
}, source = Array.from(new Array(19), (item, index) => 5 * (index + 1)), darkenSource = source.slice(0, 16).reverse();

function makeVariants(color) {
  const variants = {};
  return source.forEach(n => variants["L".concat(n)] = colorUtils.lighten(color, n)), 
  variants.base = color, darkenSource.forEach(n => variants["D".concat(n)] = colorUtils.darken(color, n)), 
  source.forEach(n => variants["A".concat(n)] = colorUtils.alpha(color, n / 100)), 
  variants;
}

const blue = "#2684FF", green = "#34c240", red = "#d64242", yellow = "#fa9f47", blues = makeVariants(blue), greens = makeVariants(green), reds = makeVariants(red), yellows = makeVariants(yellow);

var colors = _objectSpread({
  R: reds,
  G: greens,
  B: blues,
  Y: yellows
}, neutrals, {
  page: "#FAFBFC",
  text: neutrals.N90
}, intent, {
  green: green,
  red: red,
  blue: blue,
  yellow: yellow
});

const shade = alpha => "rgba(9, 30, 66, ".concat(alpha, ")"), shadows = [ "0px 2px 5px 0px ".concat(shade(.12)), "0px 5px 10px 0px ".concat(shade(.12)), "0px 6px 12px -2px ".concat(shade(.12), ", 0 0 0 1px ").concat(shade(.08)), "0px 6px 12px -2px ".concat(shade(.24), ", 0 0 0 1px ").concat(shade(.08)) ], fontFamily = '\n  -apple-system,\n  BlinkMacSystemFont,\n  "Segoe UI",\n  Roboto,\n  Helvetica,\n  Arial,\n  sans-serif,\n  "Apple Color Emoji",\n  "Segoe UI Emoji",\n  "Segoe UI Symbol"\n', borderRadius = 6, gridSize = 8, fontSize = 16, globalStyles = {
  body: {
    backgroundColor: colors.page,
    color: colors.text,
    fontFamily: fontFamily,
    fontSize: 16,
    letterSpacing: "-0.005em",
    margin: 0,
    textDecorationSkip: "ink",
    textRendering: "optimizeLegibility",
    msOverflowStyle: "-ms-autohiding-scrollbar",
    MozFontFeatureSettings: "'liga' on",
    MozOsxFontSmoothing: "grayscale",
    WebkitFontSmoothing: "antialiased"
  },
  a: {
    color: colors.primary,
    textDecoration: "none"
  },
  "a:hover": {
    textDecoration: "underline"
  }
};

exports.borderRadius = 6, exports.colors = colors, exports.fontFamily = fontFamily, 
exports.fontSize = 16, exports.globalStyles = globalStyles, exports.gridSize = 8, 
exports.shadows = shadows;
