// @arch-ui/badge/dist/badge.esm.js
import _extends from '@babel/runtime/helpers/esm/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import React, { PureComponent } from 'react';
import styled from '@emotion/styled';
import { colors } from '@arch-ui/theme';

const boldBackgroundColor = {
  default: colors.N60,
  inverted: colors.text,
  primary: colors.primary,
  created: colors.create,
  warning: colors.warning,
  danger: colors.danger
};
const boldTextColor = {
  default: 'white',
  inverted: 'white',
  primary: 'white',
  created: 'white',
  warning: 'white',
  danger: 'white'
};
const subtleBackgroundColor = {
  default: colors.N10,
  inverted: 'white',
  primary: colors.B.L85,
  created: colors.G.L85,
  warning: colors.Y.L85,
  danger: colors.R.L85
};
const subtleTextColor = {
  default: colors.N70,
  inverted: colors.text,
  primary: colors.B.D20,
  created: colors.G.D20,
  warning: colors.Y.D20,
  danger: colors.R.D20
};
const BadgeElement = styled.div((_ref) => {
  let {
    appearance,
    variant
  } = _ref;
  return {
    backgroundColor: variant === 'bold' ? boldBackgroundColor[appearance] : subtleBackgroundColor[appearance],
    borderRadius: '2em',
    boxSizing: 'border-box',
    color: variant === 'bold' ? boldTextColor[appearance] : subtleTextColor[appearance],
    display: 'inline-block',
    fontSize: 12,
    fontWeight: 500,
    lineHeight: 1,
    minWidth: '2em',
    padding: '0.25em 0.5em',
    textAlign: 'center'
  };
});
class Badge extends PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "getValue", (_ref2) => {
      let {
        value,
        max
      } = _ref2;
      if (value < 0) return '0';
      if (max > 0 && value > max) return "".concat(max, "+");
      return value;
    });
  }

  render() {
    const _this$props = this.props,
          {
      appearance,
      max,
      value,
      variant
    } = _this$props,
          rest = _objectWithoutProperties(_this$props, ["appearance", "max", "value", "variant"]);

    return React.createElement(BadgeElement, _extends({
      appearance: appearance,
      variant: variant
    }, rest), this.getValue({
      value,
      max
    }));
  }

}

_defineProperty(Badge, "defaultProps", {
  appearance: 'default',
  max: 99,
  value: 0,
  variant: 'subtle'
});

export { Badge };
