// @arch-ui/lozenge/dist/lozenge.esm.js
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import styled from '@emotion/styled';
import { colors, borderRadius, gridSize } from '@arch-ui/theme';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }
const boldBackgroundColor = {
  default: colors.N60,
  primary: colors.primary,
  danger: colors.danger,
  create: colors.create,
  warning: colors.warning
};
const boldTextColor = {
  default: 'white',
  primary: 'white',
  danger: 'white',
  create: 'white',
  warning: 'white'
};
const subtleBackgroundColor = {
  default: colors.N05,
  primary: colors.B.L90,
  danger: colors.R.L90,
  create: colors.G.L90,
  warning: colors.Y.L90
};
const subtleBorderColor = {
  default: colors.N10,
  primary: colors.B.L80,
  danger: colors.R.L80,
  create: colors.G.L80,
  warning: colors.Y.L80
};
const subtleTextColor = {
  default: colors.N50,
  primary: colors.B.L30,
  danger: colors.R.L30,
  create: colors.G.L30,
  warning: colors.Y.L30
};
const Lozenge = styled.div((_ref) => {
  let {
    appearance,
    variant,
    crop
  } = _ref;
  const isClipRight = crop === 'right';
  const cropStyles = crop ? {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    // clip from beginning of string
    direction: isClipRight ? 'rtl' : null,
    textAlign: isClipRight ? 'left' : null
  } : null;
  return _objectSpread({
    border: '1px solid transparent',
    backgroundColor: variant === 'bold' ? boldBackgroundColor[appearance] : subtleBackgroundColor[appearance],
    borderColor: variant === 'bold' ? null : subtleBorderColor[appearance],
    boxSizing: 'border-box',
    color: variant === 'bold' ? boldTextColor[appearance] : subtleTextColor[appearance],
    borderRadius,
    display: 'inline-block',
    fontSize: '0.85em',
    maxWidth: '100%',
    minWidth: 1,
    padding: "".concat(gridSize / 2, "px ").concat(gridSize, "px")
  }, cropStyles);
});
Lozenge.defaultProps = {
  appearance: 'default',
  variant: 'subtle'
};

export { Lozenge };
