// @arch-ui/select/dist/select.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var _defineProperty = _interopDefault(require('@babel/runtime/helpers/defineProperty'));
var React = require('react');
var BaseSelect = _interopDefault(require('react-select'));
var AsyncCreatableSelect = _interopDefault(require('react-select/async-creatable'));
var AsyncSelect = _interopDefault(require('react-select/async'));
var CreatableSelect = _interopDefault(require('react-select/creatable'));
var theme = require('@arch-ui/theme');

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }
// Styled Select
// ==============================

const indicatorStyles = (provided, _ref) => {
  let {
    isDisabled,
    isFocused
  } = _ref;
  let styles = {
    color: theme.colors.N20,
    ':hover': !isDisabled && !isFocused ? {
      color: theme.colors.N40
    } : null
  };
  if (isDisabled) styles = {
    color: theme.colors.N10
  };

  if (isFocused) {
    styles = {
      color: theme.colors.N60,
      ':hover': {
        color: theme.colors.N80
      }
    };
  }

  return _objectSpread({}, provided, {}, styles);
};

const selectStyles = {
  control: (provided, _ref2) => {
    let {
      isFocused
    } = _ref2;
    const focusStyles = isFocused ? {
      borderColor: theme.colors.primary,
      boxShadow: "inset 0 1px 1px rgba(0, 0, 0, 0.075),\n      0 0 0 3px ".concat(theme.colors.B.A20),
      outline: 0
    } : null;
    return _objectSpread({}, provided, {
      backgroundColor: 'white',
      borderColor: theme.colors.N20,
      fontSize: '1rem',
      minHeight: 35,
      minWidth: '200px',
      ':hover': {
        borderColor: theme.colors.N30
      }
    }, focusStyles);
  },
  clearIndicator: indicatorStyles,
  dropdownIndicator: indicatorStyles,
  menu: provided => _objectSpread({}, provided, {
    fontSize: '0.9em'
  }),
  option: (provided, _ref3) => {
    let {
      isDisabled,
      isFocused,
      isSelected
    } = _ref3;
    let bg = 'inherit';
    if (isFocused) bg = theme.colors.B.L90;
    if (isSelected) bg = theme.colors.primary;
    let txt = 'inherit';
    if (isFocused) txt = theme.colors.primary;
    if (isSelected) txt = 'white';
    if (isDisabled) txt = theme.colors.N40;
    const cssPseudoActive = !isSelected && !isDisabled ? {
      backgroundColor: theme.colors.B.L80,
      color: theme.colors.B.D20
    } : {};
    return _objectSpread({}, provided, {
      fontSize: '1rem',
      backgroundColor: bg,
      color: txt,
      ':active': cssPseudoActive
    });
  },
  menuPortal: provided => _objectSpread({}, provided, {
    zIndex: 3
  })
};

const getSelectVariant = (_ref4) => {
  let {
    isAsync,
    isCreatable
  } = _ref4;

  if (isAsync && isCreatable) {
    return AsyncCreatableSelect;
  }

  if (isAsync) {
    return AsyncSelect;
  }

  if (isCreatable) {
    return CreatableSelect;
  }

  return BaseSelect;
};

const Select = (_ref5) => {
  let {
    isAsync,
    isCreatable,
    innerRef,
    styles
  } = _ref5,
      props = _objectWithoutProperties(_ref5, ["isAsync", "isCreatable", "innerRef", "styles"]);

  const ReactSelect = getSelectVariant({
    isAsync,
    isCreatable
  });
  return React.createElement(ReactSelect, _extends({
    ref: innerRef,
    styles: React.useMemo(() => _objectSpread({}, selectStyles, {}, styles), [styles])
  }, props));
};

exports.default = Select;
