// @arch-ui/select/dist/select.cjs.prod.js
"use strict";

function _interopDefault(ex) {
  return ex && "object" == typeof ex && "default" in ex ? ex.default : ex;
}

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var _extends = _interopDefault(require("@babel/runtime/helpers/extends")), _objectWithoutProperties = _interopDefault(require("@babel/runtime/helpers/objectWithoutProperties")), _defineProperty = _interopDefault(require("@babel/runtime/helpers/defineProperty")), React = require("react"), BaseSelect = _interopDefault(require("react-select")), AsyncCreatableSelect = _interopDefault(require("react-select/async-creatable")), AsyncSelect = _interopDefault(require("react-select/async")), CreatableSelect = _interopDefault(require("react-select/creatable")), theme = require("@arch-ui/theme");

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter((function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    }))), keys.push.apply(keys, symbols);
  }
  return keys;
}

function _objectSpread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = null != arguments[i] ? arguments[i] : {};
    i % 2 ? ownKeys(Object(source), !0).forEach((function(key) {
      _defineProperty(target, key, source[key]);
    })) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach((function(key) {
      Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
    }));
  }
  return target;
}

const indicatorStyles = (provided, _ref) => {
  let {isDisabled: isDisabled, isFocused: isFocused} = _ref, styles = {
    color: theme.colors.N20,
    ":hover": isDisabled || isFocused ? null : {
      color: theme.colors.N40
    }
  };
  return isDisabled && (styles = {
    color: theme.colors.N10
  }), isFocused && (styles = {
    color: theme.colors.N60,
    ":hover": {
      color: theme.colors.N80
    }
  }), _objectSpread({}, provided, {}, styles);
}, selectStyles = {
  control: (provided, _ref2) => {
    let {isFocused: isFocused} = _ref2;
    const focusStyles = isFocused ? {
      borderColor: theme.colors.primary,
      boxShadow: "inset 0 1px 1px rgba(0, 0, 0, 0.075),\n      0 0 0 3px ".concat(theme.colors.B.A20),
      outline: 0
    } : null;
    return _objectSpread({}, provided, {
      backgroundColor: "white",
      borderColor: theme.colors.N20,
      fontSize: "1rem",
      minHeight: 35,
      minWidth: "200px",
      ":hover": {
        borderColor: theme.colors.N30
      }
    }, focusStyles);
  },
  clearIndicator: indicatorStyles,
  dropdownIndicator: indicatorStyles,
  menu: provided => _objectSpread({}, provided, {
    fontSize: "0.9em"
  }),
  option: (provided, _ref3) => {
    let {isDisabled: isDisabled, isFocused: isFocused, isSelected: isSelected} = _ref3, bg = "inherit";
    isFocused && (bg = theme.colors.B.L90), isSelected && (bg = theme.colors.primary);
    let txt = "inherit";
    return isFocused && (txt = theme.colors.primary), isSelected && (txt = "white"), 
    isDisabled && (txt = theme.colors.N40), _objectSpread({}, provided, {
      fontSize: "1rem",
      backgroundColor: bg,
      color: txt,
      ":active": isSelected || isDisabled ? {} : {
        backgroundColor: theme.colors.B.L80,
        color: theme.colors.B.D20
      }
    });
  },
  menuPortal: provided => _objectSpread({}, provided, {
    zIndex: 3
  })
}, getSelectVariant = _ref4 => {
  let {isAsync: isAsync, isCreatable: isCreatable} = _ref4;
  return isAsync && isCreatable ? AsyncCreatableSelect : isAsync ? AsyncSelect : isCreatable ? CreatableSelect : BaseSelect;
}, Select = _ref5 => {
  let {isAsync: isAsync, isCreatable: isCreatable, innerRef: innerRef, styles: styles} = _ref5, props = _objectWithoutProperties(_ref5, [ "isAsync", "isCreatable", "innerRef", "styles" ]);
  const ReactSelect = getSelectVariant({
    isAsync: isAsync,
    isCreatable: isCreatable
  });
  return React.createElement(ReactSelect, _extends({
    ref: innerRef,
    styles: React.useMemo(() => _objectSpread({}, selectStyles, {}, styles), [ styles ])
  }, props));
};

exports.default = Select;
