// @arch-ui/common/dist/common.cjs.prod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var theme = require("@arch-ui/theme");

const XL_DEVICE = 1199, LG_DEVICE = 991, MD_DEVICE = 768, SM_DEVICE = 575, smOnly = "@media (max-width: ".concat(575, "px)"), mdOnly = "@media (min-width: ".concat(768, "px) and (max-width: ").concat(991, "px)"), lgOnly = "@media (min-width: ".concat(991, "px) and (max-width: ").concat(1199, "px)"), xlOnly = "@media (min-width: ".concat(1199, "px)"), mdUp = "@media (min-width: ".concat(768, "px)"), lgUp = "@media (min-width: ".concat(991, "px)"), mdDown = "@media (max-width: ".concat(768, "px)"), lgDown = "@media (max-width: ".concat(991, "px)"), xlDown = "@media (max-width: ".concat(1199, "px)");

var mediaQueries = Object.freeze({
  __proto__: null,
  smOnly: smOnly,
  mdOnly: mdOnly,
  lgOnly: lgOnly,
  xlOnly: xlOnly,
  mdUp: mdUp,
  lgUp: lgUp,
  mdDown: mdDown,
  lgDown: lgDown,
  xlDown: xlDown
});

const uniformHeight = {
  appearance: "none",
  background: "none",
  border: "1px solid transparent",
  borderRadius: theme.borderRadius,
  boxSizing: "border-box",
  fontFamily: "inherit",
  fontSize: "1.0rem",
  lineHeight: "1.2rem",
  margin: 0,
  minWidth: 1,
  padding: "".concat(theme.gridSize, "px ").concat(1.5 * theme.gridSize, "px"),
  transition: "box-shadow 100ms linear",
  verticalAlign: "middle",
  whiteSpace: "nowrap"
};

exports.mediaQueries = mediaQueries, exports.uniformHeight = uniformHeight;
