// @keystonejs/build-field-types/dist/entrypoint.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StrictEntrypoint = exports.Entrypoint = void 0;

var _sarcastic = _interopRequireDefault(require("sarcastic"));

var _path = _interopRequireDefault(require("path"));

var _validate = require("./validate");

var _item = require("./item");

var _resolve = _interopRequireDefault(require("resolve"));

var _constants = require("./constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*::
import type { Package } from './package'
*/
class Entrypoint extends _item.Item {
  constructor(filePath, contents, pkg) {
    super(filePath, contents);
    this.package = pkg;
  }

  get name() {
    return _path.default.join(this.package.name, _path.default.relative(this.package.directory, this.directory));
  }

  get main() {
    return (0, _sarcastic.default)(this.json.main, _sarcastic.default.maybe(_sarcastic.default.string));
  }

  set main(path) {
    this.json.main = path;
  }

  get module() {
    return (0, _sarcastic.default)(this.json.module, _sarcastic.default.maybe(_sarcastic.default.string));
  }

  set module(path) {
    this.json.module = path;
  }

  get configSource() {
    return (0, _sarcastic.default)(this._config.source, _sarcastic.default.default(_sarcastic.default.string, 'src/index'));
  }

  get source() {
    return _resolve.default.sync(_path.default.join(this.directory, this.configSource), {
      extensions: _constants.EXTENSIONS
    });
  }

  strict() {
    if (!this._strict) {
      (0, _validate.validateEntrypoint)(this, false);
      this._strict = new StrictEntrypoint(this.path, this._contents, this.package);
    }

    return this._strict;
  }

}

exports.Entrypoint = Entrypoint;

class StrictEntrypoint extends Entrypoint {
  get main() {
    return (0, _sarcastic.default)(this.json.main, _sarcastic.default.string);
  }

  set main(path) {
    this.json.main = path;
  }

  get module() {
    return (0, _sarcastic.default)(this.json.module, _sarcastic.default.string);
  }

  set module(path) {
    this.json.module = path;
  }

  updater(json) {
    super.updater(json);
    (0, _validate.validateEntrypoint)(this, false);
  }

  strict() {
    return this;
  }

}

exports.StrictEntrypoint = StrictEntrypoint;
