// @arch-ui/layout/dist/layout.cjs.prod.js
"use strict";

function _interopDefault(ex) {
  return ex && "object" == typeof ex && "default" in ex ? ex.default : ex;
}

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var _defineProperty = _interopDefault(require("@babel/runtime/helpers/defineProperty")), _extends = _interopDefault(require("@babel/runtime/helpers/extends")), _objectWithoutProperties = _interopDefault(require("@babel/runtime/helpers/objectWithoutProperties")), core = require("@emotion/core"), react = require("react"), styled = _interopDefault(require("@emotion/styled")), common = require("@arch-ui/common"), theme = require("@arch-ui/theme");

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter((function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    }))), keys.push.apply(keys, symbols);
  }
  return keys;
}

function _objectSpread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = null != arguments[i] ? arguments[i] : {};
    i % 2 ? ownKeys(Object(source), !0).forEach((function(key) {
      _defineProperty(target, key, source[key]);
    })) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach((function(key) {
      Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
    }));
  }
  return target;
}

const CONTAINER_WIDTH = 1160, CONTAINER_GUTTER = 6 * theme.gridSize, Container = styled.div(_ref => {
  let {isFullWidth: isFullWidth} = _ref;
  return {
    maxWidth: isFullWidth ? "100%" : 1160,
    paddingLeft: CONTAINER_GUTTER,
    paddingRight: CONTAINER_GUTTER,
    transition: "max-width 200ms cubic-bezier(0.2, 0, 0, 1)",
    [common.mediaQueries.smOnly]: {
      paddingLeft: 2 * theme.gridSize,
      paddingRight: 2 * theme.gridSize
    }
  };
});

function collapseBorderRadii(index, length) {
  let style;
  const isFirst = 0 === index, isLast = index === length - 1;
  return isLast && !isFirst ? style = {
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0
  } : isFirst && !isLast ? style = {
    borderTopRightRadius: 0,
    borderBottomRightRadius: 0
  } : isFirst || isLast || (style = {
    borderRadius: 0
  }), style;
}

function FlexGroupComponent(_ref2, ref) {
  let {align: align = "stretch", as: Tag = "div", children: children, growIndexes: growIndexes = [], isContiguous: isContiguous, isInline: isInline, isVertical: isVertical, justify: justify = "flex-start", wrap: wrap, spacing: spacing = theme.gridSize, stretch: stretch} = _ref2, props = _objectWithoutProperties(_ref2, [ "align", "as", "children", "growIndexes", "isContiguous", "isInline", "isVertical", "justify", "wrap", "spacing", "stretch" ]);
  const gutter = spacing / 2, length = react.Children.count(children), childArray = react.Children.toArray(children).filter(child => child);
  return core.jsx(Tag, _extends({
    css: {
      alignItems: align,
      display: isInline ? "inline-flex" : "flex",
      flexDirection: isVertical ? "column" : "row",
      flexWrap: wrap ? "wrap" : "nowrap",
      justifyContent: justify,
      marginBottom: isVertical ? -gutter : null,
      marginLeft: isContiguous || isVertical ? null : -gutter,
      marginRight: isContiguous || isVertical ? null : -gutter,
      marginTop: isVertical ? -gutter : null,
      maxWidth: isInline ? "calc(100% + ".concat(2 * gutter, "px)") : null
    },
    ref: ref
  }, props), childArray.map((child, idx) => {
    const style = isContiguous ? collapseBorderRadii(idx, length) : null, leftOffset = isContiguous && idx ? -1 : gutter, rightOffset = isContiguous ? null : gutter;
    return core.jsx("div", {
      key: child.key,
      css: {
        flex: stretch || growIndexes.includes(idx) ? 1 : null,
        marginLeft: isVertical ? null : leftOffset,
        marginRight: isVertical ? null : rightOffset,
        marginTop: isVertical ? gutter : null,
        marginBottom: isVertical ? gutter : null,
        minWidth: 0,
        "&:focus-within": {
          position: "relative"
        }
      }
    }, isContiguous ? react.cloneElement(child, {
      style: style
    }) : child);
  }));
}

const FlexGroup = react.forwardRef(FlexGroupComponent);

function formatAreas(areas) {
  return areas.map(area => '"'.concat(area, '"')).join(" ");
}

const Grid = _ref3 => {
  let {alignContent: alignContent, areas: areas, columns: columns = 12, flow: flow = "row", gap: gap = 8, justifyContent: justifyContent, minRowHeight: minRowHeight = 20, rows: rows} = _ref3, props = _objectWithoutProperties(_ref3, [ "alignContent", "areas", "columns", "flow", "gap", "justifyContent", "minRowHeight", "rows" ]);
  const templateRows = rows ? {
    gridTemplateRows: rows
  } : {}, templateAreas = areas ? {
    gridTemplateAreas: formatAreas(areas)
  } : {}, gridTemplateColumns = Number.isInteger(columns) ? "repeat(".concat(columns, ", 1fr)") : columns;
  return core.jsx("div", _extends({
    css: _objectSpread({
      display: "grid",
      gridAutoFlow: flow,
      gridAutoRows: "minmax(".concat(minRowHeight, "px, auto)"),
      gridGap: gap,
      gridTemplateColumns: gridTemplateColumns,
      justifyContent: justifyContent,
      alignContent: alignContent
    }, templateRows, {}, templateAreas)
  }, props));
}, Cell = _ref4 => {
  let {area: area, height: height = 1, left: left, top: top, width: width = 1} = _ref4, props = _objectWithoutProperties(_ref4, [ "area", "height", "left", "top", "width" ]);
  return core.jsx("div", _extends({
    css: {
      alignContent: "space-around",
      gridArea: area,
      gridColumnEnd: "span ".concat(width),
      gridColumnStart: left,
      gridRowEnd: "span ".concat(height),
      gridRowStart: top,
      height: "100%",
      minWidth: 0
    }
  }, props));
};

exports.CONTAINER_GUTTER = CONTAINER_GUTTER, exports.CONTAINER_WIDTH = 1160, exports.Cell = Cell, 
exports.Container = Container, exports.FlexGroup = FlexGroup, exports.FlexGroupComponent = FlexGroupComponent, 
exports.Grid = Grid;
