// @keystonejs/fields/dist/Controller-4f311990.esm.js
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import '@babel/runtime/helpers/esm/classCallCheck';
import 'lodash.isequal';
import FieldController from '../Controller/dist/fields.esm.js';

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    if (enumerableOnly) symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    });
    keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(Object(source), true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(Object(source)).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}

class SelectController extends FieldController {
  constructor(config) {
    const defaultValue = 'defaultValue' in config ? config.defaultValue : null;

    for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      args[_key - 1] = arguments[_key];
    }

    super(_objectSpread({}, config, {
      defaultValue
    }), ...args);

    _defineProperty(this, "getFilterGraphQL", _ref => {
      let {
        value: {
          inverted,
          options
        }
      } = _ref;
      const isMulti = options.length > 1;
      let key = this.path;

      if (isMulti && inverted) {
        key = "".concat(this.path, "_not_in");
      } else if (isMulti) {
        key = "".concat(this.path, "_in");
      } else if (inverted) {
        key = "".concat(this.path, "_not");
      }

      const value = isMulti ? options.map(x => x.value).join(',') : options[0].value;
      return "".concat(key, ": ").concat(value);
    });

    _defineProperty(this, "getFilterLabel", () =>
    /*{ value }*/
    {
      return this.label;
    });

    _defineProperty(this, "formatFilter", _ref2 => {
      let {
        value
      } = _ref2;

      if (!value.options.length) {
        return value.inverted ? "".concat(this.label, " is set") : "".concat(this.label, " has no value");
      }

      if (value.options.length > 1) {
        const values = value.options.map(i => i.label).join(', ');
        return value.inverted ? "".concat(this.label, " is not in [").concat(values, "]") : "".concat(this.label, " is in [").concat(values, "]");
      }

      const optionLabel = value.options[0].label;
      return value.inverted ? "".concat(this.label, " is not ").concat(optionLabel) : "".concat(this.label, " is ").concat(optionLabel);
    });

    _defineProperty(this, "getFilterTypes", () => [{
      type: 'is',
      label: 'Matches',
      getInitialValue: () => ({
        inverted: false,
        options: []
      })
    }]);

    this.options = this.config.options;
  }

}

export default SelectController;
