// @keystonejs/fields/dist/Field-807f67c6.esm.js
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import { jsx } from '@emotion/core';
import { FieldContainer, FieldLabel, FieldDescription, FieldInput } from '@arch-ui/fields';
import { Component } from 'react';
import { Input } from '@arch-ui/input';

class UnsplashField extends Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "onChange", event => {
      this.props.onChange(event.target.value);
    });
  }

  render() {
    const {
      autoFocus,
      field,
      errors,
      value: serverValue
    } = this.props;
    const value = serverValue || '';
    const htmlID = "ks-unsplash-".concat(field.path);
    const canRead = errors.every(error => !(error instanceof Error && error.name === 'AccessDeniedError'));
    const error = errors.find(error => error instanceof Error && error.name === 'AccessDeniedError');
    return jsx(FieldContainer, null, jsx(FieldLabel, {
      htmlFor: htmlID,
      field: field,
      errors: errors
    }), field.config.adminDoc && jsx(FieldDescription, null, field.config.adminDoc), jsx(FieldInput, null, jsx(Input, {
      autoComplete: "off",
      autoFocus: autoFocus,
      type: "text",
      value: canRead ? value : undefined,
      placeholder: canRead ? 'Unsplash Image ID' : error.message,
      onChange: this.onChange,
      id: htmlID
    })));
  }

}

export default UnsplashField;
