// @keystonejs/fields/dist/Field-4e0f37af.esm.js
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import { jsx } from '@emotion/core';
import { FieldContainer, FieldLabel, FieldDescription, FieldInput } from '@arch-ui/fields';
import { Component } from 'react';
import { CheckboxPrimitive } from '@arch-ui/controls';

class TextField extends Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "onChange", event => {
      this.props.onChange(event.target.checked);
    });
  }

  render() {
    const {
      autoFocus,
      field,
      value,
      errors
    } = this.props;
    const checked = value || false;
    const htmlID = "ks-input-".concat(field.path);
    return jsx(FieldContainer, null, jsx(FieldLabel, {
      htmlFor: htmlID,
      field: field,
      errors: errors
    }), field.config.adminDoc && jsx(FieldDescription, null, field.config.adminDoc), jsx(FieldInput, {
      css: {
        height: 35
      }
    }, jsx(CheckboxPrimitive, {
      autoFocus: autoFocus,
      checked: checked,
      onChange: this.onChange,
      id: htmlID
    })));
  }

}

export default TextField;
