// @keystonejs/fields/dist/Field-1fb916cb.esm.js
import { jsx } from '@emotion/core';
import { FieldContainer, FieldLabel, FieldDescription, FieldInput } from '@arch-ui/fields';
import { TextDayPicker } from '@arch-ui/day-picker';
import { Alert } from '@arch-ui/alert';
/** @jsx jsx */

const CalendarDayField = _ref => {
  let {
    autoFocus,
    field,
    value,
    errors,
    onChange
  } = _ref;
  const htmlID = "ks-daypicker-".concat(field.path);
  return jsx(FieldContainer, null, jsx(FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), field.config.adminDoc && jsx(FieldDescription, null, field.config.adminDoc), jsx(FieldInput, null, jsx(TextDayPicker, {
    id: htmlID,
    autoFocus: autoFocus,
    date: value,
    format: field.config.format,
    onChange: onChange
  })), errors.map(_ref2 => {
    let {
      message,
      data
    } = _ref2;
    return jsx(Alert, {
      appearance: "danger",
      key: message
    }, message, data ? " - ".concat(JSON.stringify(data)) : null);
  }));
};

export default CalendarDayField;
