// @keystonejs/fields/dist/Field-89865316.cjs.js
'use strict';

var core = require('@emotion/core');

var fields = require('@arch-ui/fields');

var dayPicker = require('@arch-ui/day-picker');

var alert = require('@arch-ui/alert');
/** @jsx jsx */


const CalendarDayField = _ref => {
  let {
    autoFocus,
    field,
    value,
    errors,
    onChange
  } = _ref;
  const htmlID = "ks-daypicker-".concat(field.path);
  return core.jsx(fields.FieldContainer, null, core.jsx(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), field.config.adminDoc && core.jsx(fields.FieldDescription, null, field.config.adminDoc), core.jsx(fields.FieldInput, null, core.jsx(dayPicker.TextDayPicker, {
    id: htmlID,
    autoFocus: autoFocus,
    date: value,
    format: field.config.format,
    onChange: onChange
  })), errors.map(_ref2 => {
    let {
      message,
      data
    } = _ref2;
    return core.jsx(alert.Alert, {
      appearance: "danger",
      key: message
    }, message, data ? " - ".concat(JSON.stringify(data)) : null);
  }));
};

exports.default = CalendarDayField;
