// @keystonejs/fields/dist/Field-bf367980.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var _defineProperty = _interopDefault(require("@babel/runtime/helpers/defineProperty"));

var fields = require('@arch-ui/fields');

var React = require('react');

var React__default = _interopDefault(React);

var input = require('@arch-ui/input');

class TextField extends React.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "onChange", event => {
      const value = event.target.value; // Similar implementation as per old Keystone version

      if (/^-?\d*\.?\d*$/.test(value)) {
        this.props.onChange(value);
      }
    });
  }

  render() {
    const {
      autoFocus,
      field,
      value,
      errors
    } = this.props;
    const htmlID = "ks-input-".concat(field.path);
    return React__default.createElement(fields.FieldContainer, null, React__default.createElement(fields.FieldLabel, {
      htmlFor: htmlID,
      field: field,
      errors: errors
    }), field.config.adminDoc && React__default.createElement(fields.FieldDescription, null, field.config.adminDoc), React__default.createElement(fields.FieldInput, null, React__default.createElement(input.Input, {
      autoComplete: "off",
      autoFocus: autoFocus,
      type: "text",
      value: value,
      onChange: this.onChange,
      id: htmlID
    })));
  }

}

exports.default = TextField;
