// @keystonejs/fields/dist/Filter-eb197252.esm.js
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import '@babel/runtime/helpers/esm/objectWithoutProperties';
import '@emotion/core';
import React__default, { Component } from 'react';
import '@babel/runtime/helpers/esm/extends';
import '@arch-ui/select';
import '@babel/runtime/helpers/esm/taggedTemplateLiteral';
import '@apollo/react-hooks';
import 'graphql-tag';
import 'react-select';
import 'intersection-observer';
import { R as RelationshipSelect } from './RelationshipSelect-a88b22e2.esm.js';

const EventCatcher = _ref => {
  let {
    children
  } = _ref;
  return React__default.createElement("div", {
    onClick: e => {
      e.preventDefault();
      e.stopPropagation();
    }
  }, children);
};

class RelationshipFilterView extends Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleChange", option => {
      const {
        onChange
      } = this.props;

      if (option === null) {
        onChange(null);
      } else {
        const {
          value
        } = option;

        if (value) {
          onChange(value.id);
        }
      }
    });
  }

  render() {
    const {
      filter,
      field,
      value
    } = this.props;
    if (!filter) return null;
    const htmlID = "ks-input-".concat(field.path);
    return React__default.createElement(EventCatcher, null, React__default.createElement(RelationshipSelect, {
      field: field,
      renderContext: null,
      htmlID: htmlID,
      onChange: this.handleChange,
      value: value,
      isMulti: false
    }));
  }

}

export default RelationshipFilterView;
