// @keystonejs/fields/dist/preview-7a621635.esm.js
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import { jsx } from '@emotion/core';
import _extends from '@babel/runtime/helpers/esm/extends';

const Preview = _ref => {
  let {
    data,
    originalUrl,
    fieldPath
  } = _ref,
      props = _objectWithoutProperties(_ref, ["data", "originalUrl", "fieldPath"]);

  if (!data) {
    return null;
  }

  return jsx("div", _extends({
    css: {
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'center'
    }
  }, props, {
    id: "ks-oembed-preview-".concat(fieldPath)
  }), data.html && jsx("div", {
    css: {
      minWidth: '60px',
      maxWidth: '150px',
      flexShrink: 0,
      overflow: 'hidden',
      marginRight: '0.8em'
    },
    dangerouslySetInnerHTML: {
      __html: data.html
    }
  }), jsx("div", {
    css: {
      display: 'flex',
      flexDirection: 'column',
      whiteSpace: 'nowrap',
      overflow: 'hidden'
    }
  }, jsx("span", null, data.title), jsx("a", {
    href: originalUrl,
    target: "_blank"
  }, originalUrl)));
};

export { Preview as P };
