// @keystonejs/fields/dist/preview-0c33e771.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var _objectWithoutProperties = _interopDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var core = require('@emotion/core');

var _extends = _interopDefault(require("@babel/runtime/helpers/extends"));

const Preview = _ref => {
  let {
    data,
    originalUrl,
    fieldPath
  } = _ref,
      props = _objectWithoutProperties(_ref, ["data", "originalUrl", "fieldPath"]);

  if (!data) {
    return null;
  }

  return core.jsx("div", _extends({
    css: {
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'center'
    }
  }, props, {
    id: "ks-oembed-preview-".concat(fieldPath)
  }), data.html && core.jsx("div", {
    css: {
      minWidth: '60px',
      maxWidth: '150px',
      flexShrink: 0,
      overflow: 'hidden',
      marginRight: '0.8em'
    },
    dangerouslySetInnerHTML: {
      __html: data.html
    }
  }), core.jsx("div", {
    css: {
      display: 'flex',
      flexDirection: 'column',
      whiteSpace: 'nowrap',
      overflow: 'hidden'
    }
  }, core.jsx("span", null, data.title), core.jsx("a", {
    href: originalUrl,
    target: "_blank"
  }, originalUrl)));
};

exports.Preview = Preview;
