// @arch-ui/alert/dist/alert.cjs.prod.js
"use strict";

function _interopDefault(ex) {
  return ex && "object" == typeof ex && "default" in ex ? ex.default : ex;
}

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var styled = _interopDefault(require("@emotion/styled")), theme = require("@arch-ui/theme");

const boldBackgroundColor = {
  info: theme.colors.primary,
  success: theme.colors.create,
  warning: theme.colors.warning,
  danger: theme.colors.danger
}, boldTextColor = {
  info: "white",
  success: "white",
  warning: "white",
  danger: "white"
}, subtleBackgroundColor = {
  info: theme.colors.B.L80,
  success: theme.colors.G.L80,
  warning: theme.colors.Y.L80,
  danger: theme.colors.R.L80
}, subtleTextColor = {
  info: theme.colors.B.D20,
  success: theme.colors.G.D20,
  warning: theme.colors.Y.D20,
  danger: theme.colors.R.D20
}, Alert = styled.div(_ref => {
  let {appearance: appearance, variant: variant} = _ref;
  return {
    backgroundColor: "bold" === variant ? boldBackgroundColor[appearance] : subtleBackgroundColor[appearance],
    color: "bold" === variant ? boldTextColor[appearance] : subtleTextColor[appearance],
    borderRadius: theme.borderRadius,
    display: "flex",
    fontWeight: "bold" === variant ? 500 : null,
    minWidth: 1,
    padding: "0.9em 1.2em",
    margin: "0.9em 0",
    "& a": {
      color: "bold" === variant ? boldTextColor[appearance] : subtleTextColor[appearance],
      textDecoration: "underline"
    }
  };
});

Alert.defaultProps = {
  appearance: "info",
  variant: "subtle"
}, exports.Alert = Alert;
